---
title:  "Nanopore Report Output"
author: "B. Osberg"
date:   "10/16/2017"
output:  html_document
---

The following is a nanopore processing output document. 

```{r setup, include=FALSE}
# source("https://bioconductor.org/biocLite.R")
# biocLite("GenomicAlignments")
# BiocManager::install("GenomicAlignments")
library(GenomicAlignments)
library(data.table)
library(rmarkdown)
library(rtracklayer)

# ---- next 3 lines valid only for laptop.
# setwd("/Users/blosberg/postdoc_work/Nanopore/pigx_nanopore/")
# fin_readalignment="Input/barcode05_trimmed_aligned_spliceaware_mm2.sorted.bam"
# fin_Transcript="ref/170317_Homo_sapiens.GRCh37.87.chr.gtf"

GRobj_empty=( file.info(fin_GRobj)$size == 0 )

source('./Transcripts_funcs.R')


```

To produce this report, alignment data was taken from the following file: `r fin_readalignment`
and aligned against the genome `r Genome_version`. 
Aligned reads were then compared to the reference transcriptome database stored in the following file: `r fin_Transcript`.


```{r paragraph_GRobj_empty, eval=GRobj_empty }
cat('This pipeline has the capability to analyze raw current data from the nanopore device, however, for the present data set, the pipeline was unable to find any file containing current data; as such, these results will be omitted from the report.')
```
```{r paragraph_GRobj_notempty, eval=!GRobj_empty }
cat(paste('Raw current data from the nanopore device was imported from the following GRanges object file:', fin_GRobj)) 
```

<!-------  Load alignment data and reference transcripts: -------> 

```{r import_data, echo=FALSE}
algn = readGAlignments(fin_readalignment, use.names=TRUE, param=ScanBamParam(tag="NM", what="mapq"))
algn_cigartable = cigarOpTable(cigar(algn))
gap_index = which(colnames( algn_cigartable ) =="D" )
gap_indices = c( which( colnames( algn_cigartable ) =="I" ), which(colnames( algn_cigartable ) =="D")  )

GRalgn  = as(algn, 'GRanges')
GRLalgn = as(algn, 'GRangesList')

# GRalgn   = granges(algn)
# GRLalgn  = grglist(algn)

GRalgn$num_mismatches = GRalgn$NM  - ( algn_cigartable[ , gap_indices[1]] +algn_cigartable[ , gap_indices[2]])

#ref    ==> reference set of transcript ranges
ref_GR    = rtracklayer::import.gff(fin_Transcript) 
ref_exons_GR               =  subset(ref_GR,    type == 'exon') 
ref_exons_by_transcript    =  split(ref_exons_GR, ref_exons_GR$transcript_id )

sample_read_cum_width      = sum(  width( GRLalgn ) )
ref_transcript_cum_width   = sum(  width( ref_exons_by_transcript ) )

maxfreq = max(table(GRalgn$mapq) )
```


<!---- ========================================================= ----> 
## Correspondence with reference transcriptome:

We first consider the mapping of base-called NP data with the reference genome. The mapping quality, $MAPQ$ is defined as $-10 \ {\rm log}_{10}\left[ {\rm Prob \{\, mapping\, position\, wrong\} } \right]$, with higher values implying greater confidence in the correct alignment. This quantity has a lower bound of 0, and is rounded to the nearest integer; computation allows for this quantity to be theoretically as high as 255, though all realistic data will have $MAPQ$ values significantly lower than this. A histogram of the $MAPQ$ values obtained from the present data set is provided below. 


```{r mismatches_below_above, echo=FALSE}

half_prob=-10*log10(0.5) # -- This is the MAPQ value corresponding to 50% chance of being wrong: i.e. -10*log_{10}(0.5)

below_thresh=which( GRalgn$mapq < half_prob)
above_thresh=which( GRalgn$mapq >= half_prob )

hist(GRalgn$num_mismatches[ GRalgn$mapq < half_prob ], xlab = "number of mismatches", ylab=" frequency", main ="mapq below cutoff") 
hist(GRalgn$num_mismatches[ GRalgn$mapq >= half_prob ], xlab = "number of mismatches", ylab=" frequency", main ="mapq above cutoff") 

hist( (algn_cigartable[below_thresh,1]-GRalgn[below_thresh]$num_mismatches ) /algn_cigartable[below_thresh,1],  xlab = "percentage of accurate calls", freq=FALSE, main="accuracy -mapping below threshold" )

hist( (algn_cigartable[above_thresh,1]-GRalgn[above_thresh]$num_mismatches ) /algn_cigartable[above_thresh,1],  xlab = "percentage of accurate calls", freq=FALSE, main="accuracy -mapping above threshold" )


hist( GRalgn$mapq, max(GRalgn$mapq)+1, main = "Mapping Quality Histogram", xlab = "MAPQ")
hist( GRalgn$num_mismatches, xlab = "number of mismatches", ylab=" frequency", main ="mismatches") 


lines( c(half_prob,half_prob), c(0,maxfreq) , col="red", lwd=2)
legend( "topright" ,legend="50% confident alignment",col="red", lwd=2 )

```


Note that a $MAPQ$ value of $0$ implies a low-confidence alignment. Reads without any alignment at all were omitted from further analysis. We now partition the histogram of mismatches into the sets of reads that were below and above the mapping quality threshold `r half_prob` indicated above.

```{r pair_sample_w_ref, echo=FALSE}
# N.B. GRLalgn contains all possible alignments --some of which are repeated for a single read_id
# Now check for overlaps with reference transcriptome data:
olap_pairs  = get_olap_pairs_n_scores( GRLalgn,
                                       ref_exons_by_transcript, 
                                       sample_read_cum_width, 
                                       ref_transcript_cum_width )

#-- select a unique entry for each read_id that represents the best possible pairing 
#-- among all pairs of alignments and reference transcripts-- and compile a GRlist of such best matches
olaps_scordered               = olap_pairs[ order( olap_pairs$score, decreasing=TRUE ) ]
first_occurance_bool          = !duplicated(olaps_scordered$sample_transcript_id)
best_pairs                    = olaps_scordered [first_occurance_bool ]
```

```{r further_data, echo=FALSE}
#-- Get some supplementary data:
transcript_overlaps_best              = best_pairs
transcript_overlaps_best$N_readexons  = elementNROWS( GRLalgn[ best_pairs$queryHits ] ) 
transcript_overlaps_best$cum_width    = sum( width(   GRLalgn[ best_pairs$queryHits ] ) )

GR_original_matches                   = match(transcript_overlaps_best$sample_transcript_id, names(GRalgn))
transcript_overlaps_best$mapq         = GRalgn[GR_original_matches]$mapq

Nreads      = length( unique( names(GRLalgn) )  )
Nread_noref = Nreads -  nrow( transcript_overlaps_best )
#=======================================================================
# the fraction of reads that overlap with transcripts:
# take the number of unique occurances of hits for the overlap check, and divide by the number of unique reads:
reads_hitting_transcripts = length( unique(transcript_overlaps_best$queryHits) )

# total number of reads that aligned somewhere
tot_reads = length( unique(names(algn) ) )

# the fraction of successfully-aligned reads that overlapped at least partially with a transcript from the reference data
overlap_frac = reads_hitting_transcripts/tot_reads
```

Considering correspondence between alignment data and reference transcriptome, of the `r tot_reads` reads collected, `r reads_hitting_transcripts` showed overlaps with the reference transcriptome --i.e. `r 100*signif(overlap_frac, digits = 3)`%. For these reads, we define an overlap score as follows: assuming the cumulative read length (across all exons) from sample read $i$ is $Ls_i$, and the cumulative length of the reference transcript is $Lr_j$, while the cumulative length shared by both objects in common is $Lc$, the overlap  is given by

$$ \langle Ls_i | Lr_j \rangle = \frac{2 Lc}{Ls_i+Lr_j}.$$
The score of each read $Ls_i$ is then the best possible overlap that can be obtained upon comparison to all possible reference transcripts $Lr_j$; sample reads with no overlap have score $0$, while sample reads corresponding perfectly to a reference transcriptome can have a score as high as $1$.


Below is a histogram of mapping qualities for the subset of the reads with non-zero score,  and their 'scordered' overlaps (ordered by score == 'scordered') with known transcripts from the reference database

```{r alignment_ref_overlaps, echo=FALSE}

hist(transcript_overlaps_best$mapq, max(transcript_overlaps_best$mapq)+1, main = "Mapping Quality \n(reads on reference transcriptome)", xlab = "MAPQ")
maxfreq    = max(table(transcript_overlaps_best$mapq) )
lines( c(half_prob,half_prob), c(0,maxfreq) , col="red", lwd=2)
legend( "topright" ,legend="50% confident alignment",col="red", lwd=2 )

plot(   c(transcript_overlaps_best$score, rep(0,Nread_noref) ), 
        xlab= "index", 
        ylab="overlap score" , 
        main="Read transcript overlap \n with reference (scordered)" )
```

<!---- collect info on number of exons in each transcript: ----> 
## Exon number:
We consider the number of distinct exons from each read, and how this complexity impacts the score of the consistency between these reads and the reference trancsriptome.

Below is a barplot of the distribution of exon numbers among the read transcripts, followed by plot of the mean overlap score for each group against exon number.


```{r exon_num_dist, echo=FALSE}
M=max(transcript_overlaps_best$N_readexons)
histcount = matrix(0,M,1)
mean_score_v_Nex=matrix(0,M,1)

meanmapq_v_Nex=matrix(0,M,1) #-- mean mapping quality as a funtion of exon number
fracmapq_v_Nex=matrix(0,M,1) #-- fraction of reads above 50% alignment confidence (as a function of # exons.)

histcount = table( transcript_overlaps_best$N_readexons )
  
for ( i in c(1:M) )
  {
  
  mean_score_v_Nex[i]  = mean( 
                             transcript_overlaps_best$score[ which (
                                     transcript_overlaps_best$N_readexons == i ) ] )

    
  meanmapq_v_Nex[i]  =  mean( transcript_overlaps_best$mapq[ which(transcript_overlaps_best$N_readexons == i ) ] )

  #take fraction of reads of Nexon==i that have mapping quality >50% divided by the total number of such reads
  fracmapq_v_Nex[i]  = ( length( which( transcript_overlaps_best$mapq[ which(transcript_overlaps_best$N_readexons == i )  ] > half_prob) ) )  / length( which(transcript_overlaps_best$N_readexons == i ) ) 
  
  }

# par(mfrow=c(2,1)) 
barplot( log10( table( transcript_overlaps_best$N_readexons ) ) , 
         xlab = "", 
         ylab=expression('Log'[10]*'[freq.]') 
         )

par(mfrow=c(1,1)) 
# plot( meanmapq_v_Nex , 
#       ylim = c(0, max(meanmapq_v_Nex) ),
#       xlab = "number of exons", 
#       ylab = "mean MAPQ",
#       main = "mapping quality"
#     ) 
plot(fracmapq_v_Nex, 
     ylim = c(0, 1),
     col  ="red", 
     lwd=2,
     xlab = "number of exons",
     ylab = "mapped to ref. genome with > 50% confidence"
     )
legend( "bottomright" ,legend=" >50% confidence",col="red", lwd=2 )
```

Again, as a function of number of exons, we consider the 
```{r exon_num_and_scores, echo=FALSE}
plot( 
      mean_score_v_Nex , 
      ylim = c(0,1),
      xlab = "number of exons", 
      ylab = "Transcriptome overlap score",
      main = "olap score vs. number of exons"
    ) 
```         


<!----------------------------------------------------------------------->
## Novel Transcripts

No reference set can account for all possible transcripts throughout the genome. Thus, we consider reads that had no overlap with the reference database (i.e. 'novel'), and the frequency with which such regions were observed. A novel region with only a few reads may be dismissed as an aberration, whereas a novel region that is associated with many different reads is likely to indicate a more important transcript that the reference has omitted.


```{r Novel_transcript_info, echo=FALSE}
#-- find the reads that have no hits with existing transcripts
reads_no_reftrans_hits =  which( is.na(match( names(GRLalgn), transcript_overlaps_best$sample_transcript_id)) )

#-- take the subset of reads that had no hits:
GRLalgn_norefhit= GRLalgn[ reads_no_reftrans_hits ]

#-- reduce:
algn_norefhit_reduced = reduce( unlist( GRLalgn_norefhit) )

#-- look for histogram of overlaps:
hitcount_pairs  = findOverlaps( algn_norefhit_reduced, GRLalgn_norefhit) 

#--count how many times each reduced GRange was hit:
algn_norefhit_reduced$hits = table( queryHits(hitcount_pairs ) )

maxindex = length( algn_norefhit_reduced)

#-- filter only the ranges with significant coverage:
algn_norefhit_reduced_sighit =  algn_norefhit_reduced[ which( algn_norefhit_reduced$hits > readcov_THRESH )   ]

#-- take the subset with significant hits:
algn_NOVEL_reduced = algn_norefhit_reduced[ which( algn_norefhit_reduced$hits > readcov_THRESH ) ]

#-- now take the overlaps of the original GRanges list (without hits) against the reduced set
novel_pairs     = findOverlaps( GRLalgn_norefhit , algn_NOVEL_reduced )

```

N.B. THE METHODOLOGY FOR THIS SECTION NEEDS TO BE GENERALIZED: need to come up with a way to choose the read-coverage threshold (red horizontal line) in an objective, reproducible way.
Below is a plot of the observed number of hits across the regions that were not represented in the reference transcriptome. 
The red line delineates read hits that are considered significant enough to be treated as 'novel' transcripts, using a significance threshold of `r readcov_THRESH`.

```{r unreffed_reads_significance, echo=FALSE}

yplotmax = max(algn_norefhit_reduced$hits)
plot( algn_norefhit_reduced$hits, 
      ylim=c(0,yplotmax), 
      axes=FALSE, 
      ylab="coverage", 
      xlab="index")
# points(maxindex,yplotmax, axes=TRUE)
lines( c(1, length( algn_norefhit_reduced)),c(readcov_THRESH,readcov_THRESH), 
       col="red", 
       lwd=2 )

```

<!----------- COLLECT FINAL NOVEL TRANSCRIPTS GRLIST --------------------->

These novel and significant transcripts are summarized in the following set of GRanges:
```{r Final_novels, echo=TRUE}
NOVEL  = GRLalgn_norefhit[ queryHits( novel_pairs ) ]
```

In total, there were `r length( reduce(unlist(NOVEL) ) )` such locii identified from `r length( unique(names(NOVEL) ))` distinct reads. 
```{r print_novels_to_screen, echo=TRUE}
NOVEL
```
